docker_build.sh

```diff
@@ -1,7 +1,14 @@
 #!/bin/sh
 set -ex
+
 echo "Building the docker image"
 docker build -t wasi-sdk-builder:latest .
+
+if ! git config safe.directory | grep --quiet /workspace; then
+    echo "Setting up Git safe directory"
+    git config --global --add safe.directory /workspace
+fi
```

Comment:
why is this needed?

Reply:
The original had it:

https://github.com/WebAssembly/wasi-sdk/blob/317548590b402d622194416094fe0e392b725fe2/version.sh#L3

I didn't know why, but now I do: basically when running any `git` command inside a Docker container the user is different than usual so Git thinks something fishy is going on. Git will print something like:

```
fatal: detected dubious ownership in repository at '/workspace'
To add an exception for this directory, call:

        git config --global --add safe.directory /workspace
```

So the original author of that script inserted the line Git suggests into the `version.sh` script. If you have built wasi-sdk any number of times, you may want to check your `.gitconfig` as it could be replete with these exemptions.

Reply:
So an alternative fix here would me to make sure the `/workspace` (inside the container) appeared to be own by the current user?

Is there maybe a simple way to do that?  Modifying the `.gitconfig` that lives outside of the container seem like a bit of a hack.

Reply:
Here's an excerpt from the logs. This build job continues to fail:

```
2024-03-08T23:15:33.7528610Z #8 writing image sha256:54b42a06e7203c0f48481b8ed35206e73b76632b7c8fa05f8fff21636bca69de done
2024-03-08T23:15:33.7531519Z #8 naming to docker.io/library/wasi-sdk-builder:latest done
2024-03-08T23:15:33.7532287Z #8 DONE 2.4s
2024-03-08T23:15:33.7574889Z + git config safe.directory
2024-03-08T23:15:33.7576131Z + grep --quiet /workspace
2024-03-08T23:15:33.7591487Z + echo Setting up Git safe directory
2024-03-08T23:15:33.7592129Z Setting up Git safe directory
2024-03-08T23:15:33.7592946Z + git config --global --add safe.directory /workspace
2024-03-08T23:15:33.7605229Z + echo Building the package in docker image
2024-03-08T23:15:33.7605980Z + mkdir -p /home/runner/.ccache
2024-03-08T23:15:33.7606384Z Building the package in docker image
2024-03-08T23:15:33.7617564Z + docker run --rm -v /home/runner/work/wasi-sdk/wasi-sdk:/workspace -v /home/runner/.ccache:/root/.ccache -e NINJA_FLAGS=-v --workdir /workspace --tmpfs /tmp:exec wasi-sdk-builder:latest make package LLVM_CMAKE_FLAGS=-DLLVM_CCACHE_BUILD=ON
2024-03-08T23:15:34.0117006Z fatal: detected dubious ownership in repository at '/workspace'
2024-03-08T23:15:34.0118288Z To add an exception for this directory, call:
2024-03-08T23:15:34.0118757Z 
2024-03-08T23:15:34.0119150Z 	git config --global --add safe.directory /workspace
2024-03-08T23:15:34.0124011Z Traceback (most recent call last):
2024-03-08T23:15:34.0124634Z   File "./version.py", line 121, in <module>
2024-03-08T23:15:34.0125270Z     main(args.action, args.llvm_dir)
2024-03-08T23:15:34.0125825Z   File "./version.py", line 92, in main
2024-03-08T23:15:34.0126368Z     print(git_version())
2024-03-08T23:15:34.0126865Z   File "./version.py", line 63, in git_version
2024-03-08T23:15:34.0127760Z     '--match=wasi-sdk-*', '--dirty=+m', f'--abbrev={GIT_REF_LEN}'])
2024-03-08T23:15:34.0128517Z   File "./version.py", line 18, in exec
2024-03-08T23:15:34.0128924Z     universal_newlines=True, check=True, cwd=cwd)
2024-03-08T23:15:34.0129390Z   File "/usr/lib/python3.6/subprocess.py", line 438, in run
2024-03-08T23:15:34.0129824Z     output=stdout, stderr=stderr)
2024-03-08T23:15:34.0130809Z subprocess.CalledProcessError: Command '['git', 'describe', '--long', '--candidates=999', '--match=wasi-sdk-*', '--dirty=+m', '--abbrev=12']' returned non-zero exit status 128.
2024-03-08T23:15:34.0746203Z mkdir -p build/llvm
```

Maybe for some reason the command must run inside the Docker image?!

Reply:
the problem there is the `--global` in the git config command.  that will write to your `$HOME/.gitconfig` which is not inside the container image.   Removing the `--global` might help?   (But we have more than one repository here so it might need to be run on each repo)

Reply:
Maybe better to look into UID mapping to avoid the git warning.

Reply:
Ok, I looked at this again and tried what I think you meant by user mapping in [91c48f0](https://github.com/WebAssembly/wasi-sdk/pull/392/commits/91c48f0982a31d5988b73f3a822c44d27cbc3a6f). Basically what that means is that `docker_build.sh` will now run something like: `docker run --rm --user 1000:1000 -v /home/abrown/Code/wasi-sdk:/workspace:Z -v /home/abrown/.ccache:/root/.ccache:Z -e NINJA_FLAGS=-v --workdir /workspace --tmpfs /tmp:exec wasi-sdk-builder:latest make package LLVM_CMAKE_FLAGS=-DLLVM_CCACHE_BUILD=ON` (note the additional `:Z` and the `--user` flag). But now I get a different error related to the ccache ([here](https://github.com/WebAssembly/wasi-sdk/actions/runs/8302548419/job/22725057734?pr=392#step:10:1107):

```
ccache: error: Failed to create directory //.ccache/tmp: Permission denied
```

Reply:
Well, it looks like ce47f32 is enough to do the user mapping @sbc100 recommended. This essentially supercedes #395, allowing me at least to run the `docker_build.sh` script locally. I am slightly concerned that the Docker images stored by GitHub which have a few downloads each will now have an imprinted UID from GitHub CI; does anyone care about that?